core/src/main/java/io/spine/core/EventContextMixin.java

```diff
@@ -89,6 +90,39 @@ default ActorContext actorContext() {
         return actorContext;
     }
 
+    /**
+     * Obtains the ID of the first signal in a chain.
+     *
+     * <p>The root message is typically a {@code Command}. It can be an {@code Event} if the event
+     * was emitted by an actor directly, i.e. it was imported into the system.
+     *
+     * <p>If the associated message itself is the root of its chain, the ID cannot be assembled and
+     * thus an {@code Optional.empty()} is returned.
+     *
+     * @see Event#rootMessage()
+     */
+    @SuppressWarnings("deprecation") // For backward compatibility.
+    default Optional<MessageId> rootMessage() {
+        EventContext.OriginCase origin = getOriginCase();
+        switch (origin) {
+            case PAST_MESSAGE:
+                return Optional.of(getPastMessage().root());
+            case IMPORT_CONTEXT:
+                return Optional.empty();
```

Comment:
I docs you mention that the root can be an `event` if the event was imported into the system, but here, we return `empty` for imported origins.